test/scenarios.py

```diff
@@ -1231,9 +1231,124 @@
+                                ),
+                            },
+                        )
+                    ),
+                    action.Wait(event.FinishTransferCanceled(0, True)),
+                    action.Download(
+                        0,
+                        FILES["testfile-big"].id,
+                        "/tmp/received",
+                    ),
+                    action.Wait(
+                        event.FinishFileFailed(
+                            0, FILES["testfile-big"].id, Error.BAD_TRANSFER
+                        )
+                    ),
+                    action.NoEvent(),
+                    action.Stop(),
+                ]
+            ),
+        },
+    ),
+    Scenario(
+        "scenario4-12",
+        "Send a request with one file to a peer that's offline. Wait for Queued event and cancel the transfer from the sender side. Expect no events on the receiver once it comes online",
+        {
+            "ren": ActionList(
+                [
+                    action.Start("172.20.0.5"),
+                    action.NewTransfer("172.20.0.15", ["/tmp/testfile-big"]),
+                    action.Wait(
+                        event.Queued(
+                            0,
+                            {
+                                event.File(
+                                    FILES["testfile-big"].id, "testfile-big", 10485760
+                                ),
+                            },
+                        )
+                    ),
+                    action.CancelTransferRequest(0),
+                    action.Wait(event.FinishTransferCanceled(0, False)),
+                    action.NoEvent(),
+                    action.Stop(),
+                ]
+            ),
+            "stimpy": ActionList(
+                [
+                    action.Sleep(8),
+                    action.Start("172.20.0.15"),
+                    action.NoEvent(),
+                    action.Stop(),
+                ]
+            ),
+        },
+    ),
+    Scenario(
+        "scenario4-13",
+        "Send a request with one file to a peer that's offline. Cancel the transfer from the sender side immediately. Expect no events on the receiver once it comes online",
+        {
+            "ren": ActionList(
+                [
+                    action.Start("172.20.0.5"),
+                    action.NewTransfer("172.20.0.15", ["/tmp/testfile-big"]),
+                    action.CancelTransferRequest(0),
+                    action.Wait(event.FinishTransferCanceled(0, False)),
+                    action.NoEvent(),
+                    action.Stop(),
+                ]
+            ),
+            "stimpy": ActionList(
+                [
+                    action.Sleep(8),
+                    action.Start("172.20.0.15"),
+                    action.NoEvent(),
+                    action.Stop(),
+                ]
+            ),
+        },
+    ),
```

Comment:
This test fails on the sender side with a `BadTransfer` :
```
test-ren-1     | 2023-07-12 15:23:23,786:global:DEBUG:Action threw an exception: Unexpected event:
test-ren-1     | FinishFailedTransfer(transfer=614fc4f9-9482-47d3-a2c8-4c2f22e8975a (slot: 0), status=7, os_err=None)
test-ren-1     | while looking for:
test-ren-1     | FinishTransferCanceled(transfer=614fc4f9-9482-47d3-a2c8-4c2f22e8975a (slot: 0), by_peer=False)
```

Which is not what happened.

@matszczygiel it looks like the logic depends on the connection to the peer. Could you take a look into these testcases?

Reply:
We definitely have a bug here, I discovered it recently, which prevents from canceling when the peer is offline